osu.Game/Screens/Play/ReplayPlayerLoader.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Logging;
+using osu.Framework.Screens;
+using osu.Game.Beatmaps;
+using osu.Game.Online.API;
+using osu.Game.Online.API.Requests;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Scoring;
+using osu.Game.Scoring.Legacy;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace osu.Game.Screens.Play
+{
+    public class ReplayPlayerLoader : PlayerLoader
+    {
+        private readonly ScoreInfo score;
+
+        private Score replayScore;
+
+        private readonly Func<Score, Player> createPlayerWithReplay;
+
+        [Resolved]
+        private IAPIProvider api { get; set; }
+
+        [Resolved]
+        private RulesetStore rulesets { get; set; }
+
+        [Resolved]
+        private BeatmapManager beatmaps { get; set; }
+
+        public ReplayPlayerLoader(ScoreInfo score, Func<Score, Player> createPlayerWithReplay)
+            : base(null)
+        {
+            this.score = score;
+            this.createPlayerWithReplay = createPlayerWithReplay;
+        }
+
+        protected virtual Score CreateReplayScore(string filename)
+        {
+            using (var stream = new FileStream(filename, FileMode.Open))
+            {
+                return new Score
+                {
+                    ScoreInfo = score,
+                    Replay = new DatabasedLegacyScoreParser(rulesets, beatmaps).Parse(stream).Replay,
+                };
+            }
+        }
+
+        protected void LoadReplay(string filename, Action<Player> onLoad)
+        {
+            replayScore = CreateReplayScore(filename);
+
+            var player = createPlayerWithReplay(replayScore);
+
+            LoadComponentAsync(player, onLoad);
+        }
+
+        protected override IReadOnlyList<Mod> DisplayMods => score.Mods;
+
+        protected override Task CreatePlayerLoadTask(Action<Player> onLoad)
+        {
+            return Task.Run(() =>
```

Comment:
this shouldbe done in `ReplayPlayer`, not here. `PlayerLoader` already supports async loading of the player. there's likely little reason to over-complicate the class to allow for this.

Reply:
i was aiming for leaving `ReplayPlayer` as a class that just plays a given score with a replay, and keep the download logic external to it. this was because of the need to perform an async pre-load operation before the loading of player itself.

Reply:
what is the reason for the pre-load being required?

Reply:
the second sentence was unrelated to the first, sorry.

i meant this code exists because of the need to download the replay before beginning the player load since `ReplayPlayer` needs to be instantiated with the replay.

Reply:
I see. I'm hesitant to move actual loading code into the Loader classes, as these are not supposed to be doing loading – only display the progress of loading.

Another option is to move the replay download code into a button class (see `DownloadButton`). A quick implementation would be do just make a local button; a more thorough one would be to move the replay download code to `ScoreManager`, make a common interface like `IModelDownloader`, and making `DownloadTrackingComposite` handle this.

If you want to get replay loading in quickly, I'm fine with you not doing the second part of the above. Up to you.

Either way though, I believe you will want to move the download code into `ScoreManager` and use `OsuGame.PresentScore` process like dragged replays to. Best to keep things simple and using the same flow.